### PR TITLE
EI-555 Add data onto the Application Submitted event

### PR DESCRIPTION
**Pull request by ruslanss**
# Conflicts:
#	src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/Steps/ApprenticeshipIncentiveCreatedSteps.cs
#	src/SFA.DAS.EmployerIncentives.Commands.Types/ApprenticeshipIncentive/CreateIncentiveCommand.cs
#	src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/CreateIncentive/CreateIncentiveCommandHandler.cs
#	src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/Events/Submitted.cs
#	src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs
#	src/SFA.DAS.EmployerIncentives.Events/IncentiveApplications/SubmittedHandler.cs
#	src/SFA.DAS.EmployerIncentives.Functions.TestConsole/NServiceBusConsole.cs
#	src/tests/SFA.DAS.EmployerIncentives.Commands.UnitTests/ApprenticeshipIncentive/CreateIncentive/Handlers/WhenHandlingCreateIncentiveCommand.cs

### CODE REVIEW

File: src/SFA.DAS.EmployerIncentives.Commands.Types/ApprenticeshipIncentive/CreateIncentiveCommand.cs [hunk 1]
```diff
@@ -1,37 +1,64 @@
...
 {
     public class CreateIncentiveCommand : DomainCommand, ILockIdentifier, ILogWriter
     {
-        public long AccountId { get; private set; }
-        public Guid IncentiveApplicationId { get; private set; }
-        public string LockId { get => $"{nameof(Account)}_{AccountId}"; }
+        public long AccountId { get; }
+        public long AccountLegalEntityId { get; }
+        public string LockId => $"{nameof(Account)}_{AccountId}";
+        public List<IncentiveApprenticeship> Apprenticeships { get; set; }
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Commands.Types/ApprenticeshipIncentive/CreateIncentiveCommand.cs:16]: Shouldn't there be one apprenticeship per command?  This command could be quite big.

File: src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs [hunk 2]
```diff
@@ -49,11 +49,13 @@ public void Submit(DateTime submittedAt, string submittedByEmail, string submitt
...
-                 SubmittedAt = submittedAt,
-                 SubmittedBy = submittedByName,
-                 SubmittedByEmail = submittedByEmail
+                AccountId = AccountId,
+                IncentiveApplicationId = Id,
+                SubmittedAt = submittedAt,
+                SubmittedBy = submittedByName,
+                SubmittedByEmail = submittedByEmail,
+                AccountLegalEntityId = Model.AccountLegalEntityId,
+                Apprenticeships = Model.ApprenticeshipModels,
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs:58]: This shouldn't be put on the domain event here.  The Submit method only knows about the submittedAt, submittedBy and submittedByEmail information.  You are coupling this domain event to another client.  That is what the event handlers are for.  AccountLegalEntityId  and Apprenticeships should't be in this domain event you have coupled a use case from the other domain to this aggregate.

File: src/SFA.DAS.EmployerIncentives.Events/IncentiveApplications/SubmittedHandler.cs [hunk 3]
```diff
@@ -20,10 +23,25 @@ public Task Handle(Submitted @event, CancellationToken cancellationToken = defau
...
+                @event.AccountLegalEntityId,
+                MapForIncentive(@event.Apprenticeships)
                 );
 
             return _commandPublisher.Publish(command);
         }
+
+        private static List<CreateIncentiveCommand.IncentiveApprenticeship> MapForIncentive(IEnumerable<ApprenticeshipModel> apprenticeships)
+        {
+            return apprenticeships.Select(a => new CreateIncentiveCommand.IncentiveApprenticeship(
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Events/IncentiveApplications/SubmittedHandler.cs:35]: This is where the apprenticeships for the submitted application should be looked up and a command per apprenticeship created.  Then this has the single responsibility of converting a submitted event into a CreateIncentiveCommand.

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs (hunk 2 above)
Comment (MichaelYoung1981) [reply src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs:58]: I don't think that's necessarily correct that it's coupling the event to the client.  It isn't unreasonable to expose what has been submitted on a submitted event.

File: src/SFA.DAS.EmployerIncentives.Events/IncentiveApplications/SubmittedHandler.cs (hunk 3 above)
Comment (MichaelYoung1981) [reply src/SFA.DAS.EmployerIncentives.Events/IncentiveApplications/SubmittedHandler.cs:35]: Which bounded context does this class form part of?  Hard to tell from our structure as everything is intermingled.

File: src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs (hunk 2 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs:58]: Why not just raise the event with the Model in that case? e.g.

 AddEvent(new Submitted   { IncentiveApplication= Model} ?

Then you get all the data in the event.

File: src/SFA.DAS.EmployerIncentives.Events/IncentiveApplications/SubmittedHandler.cs (hunk 3 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Events/IncentiveApplications/SubmittedHandler.cs:35]: The folder structure gives the scope.  So this is in the IncentiveApplications folder so its part of that bounded context.  We could move into separate projects to make clearer perhaps?

File: src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs (hunk 2 above)
Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Domain/IncentiveApplications/IncentiveApplication.cs:58]: sounds good to me :)

File: src/SFA.DAS.EmployerIncentives.Commands.Types/ApprenticeshipIncentive/CreateIncentiveCommand.cs (hunk 1 above)
Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Commands.Types/ApprenticeshipIncentive/CreateIncentiveCommand.cs:16]: agreed

**Review by David-B-Read — APPROVED** [review]

**Review by Stephek — APPROVED** [review]
Nice work!

**Pull request merged**